transactor/src/main/scala/io/mediachain/transactor/TypeSerialization.scala

```diff
@@ -0,0 +1,243 @@
+        case _ => Xor.left(UnknownObjectType(typeName))
+      }
+    } yield entry
+
+
+  def canonicalEntryFromCMap(cMap: CMap)
+  : Xor[DeserializationError, CanonicalEntry] =
+    for {
+      index <- getRequired[CInt](cMap, "index").map(_.num)
+      ref <- getRequiredReference(cMap, "ref")
+    } yield CanonicalEntry(index, ref)
+
+
+  def chainEntryFromCMap(cMap: CMap)
+  : Xor[DeserializationError, ChainEntry] =
+    for {
+      index <- getRequired[CInt](cMap, "index").map(_.num)
+      ref <- getRequiredReference(cMap, "ref")
+      chain <- getRequiredReference(cMap, "chain")
+    } yield ChainEntry(
+      index = index,
+      ref = ref,
+      chain = chain,
+      chainPrevious = getOptionalReference(cMap, "chainPrevious")
+    )
+
+  def journalEntriesFromCArray(cArray: CArray)
+  : Xor[DeserializationError, Array[JournalEntry]] = {
+    val entryCMaps = cArray.items
+      .flatMap(v => Try(v.asInstanceOf[CMap]).toOption)
+
+    val entryXors: List[Xor[DeserializationError, JournalEntry]] =
+      entryCMaps.map(journalEntryFromCMap)
+
+    val initial: Xor[DeserializationError, List[JournalEntry]] =
+      Xor.right[DeserializationError, List[JournalEntry]](List())
+
+    val entriesListXor: Xor[DeserializationError, List[JournalEntry]] =
+      entryXors.foldLeft(initial) { (accum, x) =>
+        (accum, x) match {
+          case (Xor.Left(err), _) => Xor.left(err)
+          case (_, Xor.Left(err)) => Xor.left(err)
+          case (Xor.Right(list), Xor.Right(entry)) => Xor.right(entry :: list)
+        }
+      }
+
+    entriesListXor.map(_.reverse.toArray)
+  }
+
+  def journalBlockFromCMap(cMap: CMap)
+  : Xor[DeserializationError, JournalBlock] =
+    for {
+      index <- getRequired[CInt](cMap, "index").map(_.num)
+      entriesCArray <- getRequired[CArray](cMap, "entries")
+      entries <- journalEntriesFromCArray(entriesCArray)
+    } yield
+      JournalBlock(
+        index = index,
+        chain = getOptionalReference(cMap, "chain"),
+        entries = entries
+      )
+
+  def getRequired[T <: CValue](cMap: CMap, fieldName: String)
+  : Xor[RequiredFieldNotFound, T] = Xor.fromOption(
+    cMap.getAs[T](fieldName),
+    RequiredFieldNotFound(fieldName)
+  )
+
+
+  def getRequiredReference(cMap: CMap, fieldName: String)
+  : Xor[DeserializationError, Reference] =
+    getRequired[CMap](cMap, fieldName)
+    .flatMap(referenceFromCMap)
+
+
+  def getOptionalReference(cMap: CMap, fieldName: String): Option[Reference] =
+    cMap.getAs[CMap](fieldName)
+      .flatMap(referenceFromCMap(_).toOption)
+
+
+
+  def referenceFromCMap(cMap: CMap): Xor[DeserializationError, Reference] = {
+    val linkOpt: Option[CValue] =
+      cMap.asStringKeyedMap.get("@link")
+
+    Xor.fromOption(linkOpt, ReferenceDecodingFailed("@link field not found"))
+      .flatMap(decodeLinkValue)
+  }
+
+
+  def decodeLinkValue(linkVal: CValue): Xor[DeserializationError, Reference] =
```

Comment:
Yeah, I don't like this either....  is it possible to define a static method inside a trait?  Seems like probably not, since it would really be a method on the companion object... but it would be nice to have a `fromCValue` method on the `CborSerializable` trait that does the actual unpacking.